[New Blog] Introducing the BigCodeBench Leaderboard (#2144)
huggingface/blog: leaderboard-bigcodebench.md

---
title: "BigCodeBench: Benchmarking Large Language Models on Solving Practical and Challenging Programming Tasks"

> thumbnail: 

Reviewer: Note: missing. We also need an entry in `_blog.yml`.

Author: done :)

authors:
- user: terryyz
  guest: true
  org: bigcode
- user: ganler
  guest: true
  org: bigcode
- user: SivilTaram

>   guest: true

Reviewer: Any affiliation for this author?

Author: Oh, missed. Thanks for the note.

- user: huybery

>   guest: true

Reviewer: ```suggestion
  guest: true
  org: bigcode
```

maybe?

- user: Muennighoff
  guest: true
  org: bigcode
- user: dpfried
  guest: true
  org: bigcode
- user: harmdevries
  guest: true
  org: bigcode
- user: lvwerra
- user: clefourrier
---

# BigCodeBench: Benchmarking Large Language Models on Solving Practical and Challenging Programming Tasks

> [HumanEval](https://github.com/openai/human-eval) is a widely used benchmark for evaluating large language models (LLMs) on code generation tasks. A main reason for its widespread use is that it is easy to evaluate a compact function-level code snippet. However, there are growing concerns about the effectiveness of HumanEval in evaluating programming capabilities of LLMs. The main concern is that the tasks in HumanEval are too simple and may not be representative of real-world programming tasks. Compared to the algorithm-oriented tasks in HumanEval, real-world software developement is often involved with diverse libraries and function calls. Futhermore, LLMs' performance on HumanEval is subject to [contamination and overfitting issues](https://arxiv.org/abs/2403.07974), making it less reliable for evaluating the generalization of LLMs.

Reviewer 2: ```suggestion
[HumanEval](https://github.com/openai/human-eval) is a reference benchmark for evaluating large language models (LLMs) on code generation tasks, as it makes the evaluation of compact function-level code snippets easy. However, there are growing concerns about its effectiveness in evaluating programming capabilities of LLMs, and the main concern is that tasks in HumanEval are too simple, and may not be representative of real-world programming tasks. Compared to the algorithm-oriented tasks in HumanEval, real-world software developement is often involved with diverse libraries and function calls. Futhermore, LLMs' performance on HumanEval is subject to [contamination and overfitting issues](https://arxiv.org/abs/2403.07974), making it less reliable for evaluating the generalization of LLMs.
```

> While there have been some efforts to address these issue, they are either domain-specific, determinisitc, or agent-centric (sorry [DS-1000](https://github.com/HKUNLP/DS-1000), [ODEX](https://github.com/zorazrw/odex), and [SWE-bench](https://github.com/princeton-nlp/SWE-bench) 💔).

Reviewer: ```suggestion
While there have been some efforts to address these issues, they are either domain-specific, determinisitc, or agent-centric (sorry [DS-1000](https://github.com/HKUNLP/DS-1000), [ODEX](https://github.com/zorazrw/odex), and [SWE-bench](https://github.com/princeton-nlp/SWE-bench) 💔).
```

> The community still lacks an easy-to-use benchmark that can broadly evaluate the programming capabilities of LLMs.

Reviewer 2: ```suggestion
We feel that the community still lacks an easy-to-use benchmark that can broadly evaluate the programming capabilities of LLMs, and that's what we focused on.
```

> We are excited to announce the release of BigCodeBench, which evaluates LLMs on solving practical and challenging programming tasks without contamination. Specifically, BigCodeBench contain 1,140 function-level tasks to challenge LLMs to follow instruction and compose multiple function calls as tools from 139 libraries. To evaluate LLMs rigorously, each programming task encompasses 5.6 test cases with an average branch coverage of 99%.

Reviewer: ```suggestion
We are excited to announce the release of BigCodeBench, which evaluates LLMs on solving practical and challenging programming tasks without contamination. Specifically, BigCodeBench contains 1,140 function-level tasks to challenge LLMs to follow instructions and compose multiple function calls as tools from 139 libraries. To evaluate LLMs rigorously, each programming task encompasses 5.6 test cases with an average branch coverage of 99%.
```

Does `tools` refer to LLM agentic calls?

> Ready to deep dive into BigCodeBench? Let's get started! 🚀

Reviewer: ```suggestion
Ready to dive into BigCodeBench? Let's get started! 🚀
```

    def test_response_content(self, mock_conn):
        """ Test the content of the response. """
        mock_conn.return_value.getresponse.return_value.read.return_value = b'Expected Content'
        result = task_func('www.example.com', 443, '/content/path')
        self.assertEqual(result, 'Expected Content')

    # Mock the failed connection and assess the error handling
    @patch('socket.create_connection')
    @patch('http.client.HTTPSConnection')
    def test_ssl_handshake_error_handling(self, mock_conn, mock_socket):
        """ Test handling of SSL handshake errors. """
        mock_socket.side_effect = ssl.SSLError('SSL handshake failed')
        with self.assertRaises(ssl.SSLError):
            task_func('badssl.com', 443, '/test/path')

    # More test cases...
```
BigCodeBench features complex, user-oriented instructions for each task, including clear functionality descriptions, input/output formats, error handling, and verified interactive examples. We avoid step-by-step task instructions, believing capable LLMs should understand and solve tasks from the user's perspective in an open-ended manner. We verify specific features using test cases.

> 

Reviewer 2: this paragraph should go above the test description

## Where do the tasks come from? 🤔  

<img src="https://github.com/bigcode-bench/bigcode-bench.github.io/blob/main/asset/construct_pipeline.svg?raw=true" alt="png" style="display: block; margin-left: auto; margin-right: auto;">

We guarantee the quality of the tasks in BigCodeBench through a systematic "Human-LLM collaboration process." We start with [ODEX](https://github.com/zorazrw/odex) as the "seed dataset", which contains short but realistic human intents and corresponding Python one-liners from Stack Overflow. We use GPT-4 to expand these one-liners into comprehensive function-level tasks.

> Next, 20 human experts—most holding PhD degrees and with over 5 years of Python programming experience—voluntarily guide GPT-4 in an execution-based sandbox. They continually instruct it to refine the synthesized tasks and add test cases. The tasks and test cases are then examined in a local environment, pre-evaluated on other LLMs, and cross-checked by 7 additional human experts to ensure their quality.

Reviewer: ```suggestion
Next, 20 human experts—most holding PhD degrees and with over 5 years of Python programming experience—voluntarily guide GPT-4 in an execution-based sandbox. They continually instruct it to refine the synthesized tasks and add test cases. The tasks and test cases are then examined in a local environment, pre-evaluated on other LLMs, and cross-checked by 7 additional human experts to ensure their quality.
```

I would just say "experts" instead of mentioning the PhD degrees, may sound a bit like gatekeeping.

To help the community understand model performance on each task, we track solve rates, measured by calibrated Pass@1. On BigCodeBench-Complete, 151 tasks remain unsolved by all 71 models, while 6 tasks are completely solved. For BigCodeBench-Instruct, 281 tasks remain unsolved and 14 tasks are fully solved by all models. The significant number of unsolved tasks and the small number of fully solved tasks show that BigCodeBench is a challenging benchmark for LLMs.

## Great! So how can I evaluate my model on BigCodeBench? 🛠️

> We make BigCodeBench easily accessible to the community by providing a simple and user-friendly evaluation framework, which can be downloaded via [PyPI](https://pydigger.com/pypi/bigcodebench). The prototype of the evaluation framework is based on [EvalPlus](https://github.com/evalplus/evalplus) for the HumanEval+ and MBPP+ benchmarks. However, as our benchmark has tasks with much more diverse library dependencies than EvalPlus, we build less resource-constrained execution environment, and adapt it for `unittest` in the test harness of BigCodeBench.

Reviewer: ```suggestion
We make BigCodeBench easily accessible to the community by providing a simple and user-friendly evaluation framework, which can be downloaded via [PyPI](https://pydigger.com/pypi/bigcodebench). The prototype of the evaluation framework is based on [EvalPlus](https://github.com/evalplus/evalplus) for the HumanEval+ and MBPP+ benchmarks. However, as our benchmark has tasks with much more diverse library dependencies than EvalPlus, we build a less resource-constrained execution environment, and adapt it for `unittest` in the test harness of BigCodeBench.
```

To facilitate the evaluation, we provide pre-built Docker images for [_code generation_](https://hub.docker.com/r/terryzho/bigcodebench-generate), and [_code execution_](https://hub.docker.com/r/terryzho/bigcodebench-evaluate). Check out our [GitHub repository](https://github.com/bigcode-project/bigcodebench) to find more details on how to use the evaluation framework.

### Setup
```bash
# Install to use bigcodebench.evaluate
pip install bigcodebench --upgrade

> # If you want to use the evaluate locally, you need to install the requirements

Reviewer: ```suggestion
# If you want to evaluate locally, you need to install the requirements
```

# Install to use bigcodebench.generate
# You are strongly recommended to install the [generate] dependencies in a separate environment
pip install bigcodebench[generate] --upgrade
```

### Code Generation
You are suggested to use `flash-attn` for generating code samples.
```bash
pip install -U flash-attn
```

To generate code samples from a model, you can use the following command:
```bash
bigcodebench.generate \
    --model [model_name] \
    --subset [complete|instruct] \
    --greedy \
    --bs [bs] \
    --temperature [temp] \
    --n_samples [n_samples] \
    --resume \
    --backend [vllm|hf|openai|mistral|anthropic|google] \
    --tp [gpu_number] \
    [--trust_remote_code] \
    [--base_url [base_url]]
```

The generated code samples will be stored in a file named `[model_name]--bigcodebench-[instruct|complete]--[backend]-[temp]-[n_samples].jsonl`.

### Code Post-processing

LLM-generated text may not be compilable code as it includes natural language lines or incomplete extra code.
We provide a tool namely `bigcodebench.sanitize` to clean up the code:

```bash

> # 💡 If you want to storing calibrated codes in jsonl:

Reviewer: ```suggestion
# 💡 If you want to store calibrated code in jsonl:
```

bigcodebench.sanitize --samples samples.jsonl --calibrate
# Sanitized code will be produced to `samples-sanitized-calibrated.jsonl`

# 💡 If you do without calibration:
bigcodebench.sanitize --samples samples.jsonl
# Sanitized code will be produced to `samples-sanitized.jsonl`

# 💡 If you are storing codes in directories:
bigcodebench.sanitize --samples /path/to/vicuna-[??]b_temp_[??]
# Sanitized code will be produced to `/path/to/vicuna-[??]b_temp_[??]-sanitized`
```

### Code Evaluation

You are strongly recommended to use a sandbox such as [docker](https://docs.docker.com/get-docker/):

```bash
# Mount the current directory to the container
docker run -v $(pwd):/app bigcodebench/bigcodebench-evaluate:latest --subset [complete|instruct] --samples samples-sanitized-calibrated
# ...Or locally ⚠️
bigcodebench.evaluate --subset [complete|instruct] --samples samples-sanitized-calibrated
# ...If the ground truth is working locally (due to some flaky tests)
bigcodebench.evaluate --subset [complete|instruct] --samples samples-sanitized-calibrated --no-gt
```

## What's next?

> We share a long-term roadmap to address the limitations of BigCodeBench and sustainably build with the community. We believe that [program-aided language models](https://arxiv.org/abs/2211.10435) for task completion and reasoning offer a path toward artificial general intelligence. Our goal is to provide the community with the most open, reliable, and scalable evaluations to truly understand the fundamental capabilities of LLMs for programming and pinpoint ways to unleash their power. Specifically, we plan to enhance the following aspects of BigCodeBench:

Reviewer 2: "We believe that [program-aided language models](https://arxiv.org/abs/2211.10435) for task completion and reasoning offer a path toward artificial general intelligence." this sentence is debatable and a bit out of scope in this paragraph, I would suggest removing it

- **Multilingualism**: Currently, BigCodeBench is Python-only and cannot be easily extended to other programming languages. Since function calls are mostly language-specific, finding packages or libraries with the same functionalities in languages other than Python is challenging.

> - **Rigorousness**: While we achieve high test coverage for ground-truth solutions in BigCodeBench, it does not guarantee that *all* code solutions generated by LLMs will be correctly assessed against existing test cases. Previous works like EvalPlus have attempted to extend limited test cases by augmenting input-output pairs via LLM- and mutation-based strategies. However, adapting EvalPlus to the test harness in BigCodeBench is challenging, as it only emphasizes the input-output assertions. However, 

Reviewer: ```suggestion
- **Rigorousness**: While we achieve high test coverage for ground-truth solutions in BigCodeBench, it does not guarantee that *all* code solutions generated by LLMs will be correctly assessed against existing test cases. Previous works like EvalPlus have attempted to extend limited test cases by augmenting input-output pairs via LLM- and mutation-based strategies. However, adapting EvalPlus to the test harness in BigCodeBench is challenging, as it only emphasizes the input-output assertions.
```

Author: Changed to `However, adapting EvalPlus to the test harness in BigCodeBench is challenging. While EvalPlus only emphasizes the input-output assertions, most of test harnesses in BigCoeBench require non-trivial configurations (e.g., mock patching) to examine expected program behaviors during runtime.`

most of test harnesses in BigCoeBench require non-trivial configurations (e.g., mock patching) to examine expected program behaviors during runtime.

- **Generalization**: A key question is, "How well do the models generalize to unseen tools and tasks?" Currently, BigCodeBench covers common libraries and daily programming tasks. Benchmarking models on programming tasks that use emerging libraries like [transformers](https://github.com/huggingface/transformers) and [langchain](https://github.com/langchain-ai/langchain) would be more interesting.

- **Evolution**: Libraries can become obsolete or be updated, meaning the source code data for model training will constantly evolve. Models may not memorize function calls from deprecated library versions, posing a challenge for any tool-dependent programming benchmarks to correctly examine model capabilities without periodic updates. Another related concern is test set contamination due to evolving training data.

> - **Interaction**: Recent interest focuses on the concept of _LLMs as Agents_, which is seen as a path toward artificial general intelligence. Specifically, LLMs will be grounded in a less constrained sandbox environment, where they can interact with applications such as web browsers and terminals. This environment can help unlock capabilities like [self-debugging](https://arxiv.org/pdf/2304.05128) and [self-reflection](https://arxiv.org/abs/2303.11366).

Reviewer: I have the same comment here as Clémentine had before, in regards to the path towards AGI.

---

> title: "Introducing the BigCodeBench Leaderboard: Benchmarking Large Language Models on Solving Practical and Challenging Programming Tasks"

Reviewer 3: Simplification:  "BigCodeBench: Benchmarking Large Language Models on Solving Practical and Challenging Programming Tasks"

thumbnail: 
authors:
- user: terryyz
  guest: true
  org: bigcode
- user: ganler
  guest: true
  org: bigcode
- user: zijwang
  guest: true
  org: bigcode
- user: SivilTaram
  guest: true
- user: huybery
  guest: true
- user: Muennighoff
  guest: true
  org: bigcode
- user: dpfried
  guest: true
  org: bigcode
- user: harmdevries
  guest: true
  org: bigcode
- user: lvwerra
- user: clefourrier
---

# Introducing the BigCodeBench Leaderboard: Benchmarking Large Language Models on Solving Practical and Challenging Programming Tasks

> [HumanEval](https://github.com/openai/human-eval) is a widely used benchmark for evaluating large language models (LLMs) on code generation tasks. One main reason is that it is easy to evaluate a condense function-level code snippet. There are growing concerns about the effectiveness of HumanEval in evaluating programming capabilities of LLMs. The main concern is that the tasks in HumanEval are too simple and may not be representative of real-world programming tasks.

Reviewer 3: I think contamination is general to non-live public benchmarks (including BigCodeBench). :)
However, many training frameworks now also supports decomtaminator (https://github.com/huggingface/alignment-handbook/blob/606d2e954fd17999af40e6fb4f712055ca11b2f0/src/alignment/decontaminate.py#L67) for some datasets -- we can also build one for BigCodeBench in the future.

> [HumanEval](https://github.com/openai/human-eval) is a widely used benchmark for evaluating large language models (LLMs) on code generation tasks. One main reason is that it is easy to evaluate a condense function-level code snippet. There are growing concerns about the effectiveness of HumanEval in evaluating programming capabilities of LLMs. The main concern is that the tasks in HumanEval are too simple and may not be representative of real-world programming tasks.

Author: Agree, and we can think of another way: how about continuing to utilize LLMs to synthesize tasks like the ones in BigCodeBench but with new function calls? That is mentioned in the "What's next" section. IMO contamination is not actually a big issue if we can develop the benchmark in the long term.

> [HumanEval](https://github.com/openai/human-eval) is a widely used benchmark for evaluating large language models (LLMs) on code generation tasks. One main reason is that it is easy to evaluate a condense function-level code snippet. There are growing concerns about the effectiveness of HumanEval in evaluating programming capabilities of LLMs. The main concern is that the tasks in HumanEval are too simple and may not be representative of real-world programming tasks.

Author: We may not discuss the contamination issue here, which is not our main purpose when creating this benchmark.

For now, considering that this benchmark is actually built from scratch, contamination is quite minimal for now.

> [HumanEval](https://github.com/openai/human-eval) is a widely used benchmark for evaluating large language models (LLMs) on code generation tasks. One main reason is that it is easy to evaluate a condense function-level code snippet. There are growing concerns about the effectiveness of HumanEval in evaluating programming capabilities of LLMs. The main concern is that the tasks in HumanEval are too simple and may not be representative of real-world programming tasks.

Reviewer 2: @terryyz I actually thought that saying "HumanEval is now contaminated, hence another reason why new benchmarks are good" could have strengthened the argument (though I agree that ofc your benchmark will get contaminated at some point).

> [HumanEval](https://github.com/openai/human-eval) is a widely used benchmark for evaluating large language models (LLMs) on code generation tasks. One main reason is that it is easy to evaluate a condense function-level code snippet. There are growing concerns about the effectiveness of HumanEval in evaluating programming capabilities of LLMs. The main concern is that the tasks in HumanEval are too simple and may not be representative of real-world programming tasks.

Author: Yeah, it's a good point. I'll add that :)

> While there have been some efforts to address this issue, they are either domain-specific, solution-specific, or hard-to-deploy (sorry [DS-1000](https://github.com/HKUNLP/DS-1000), [ODEX](https://github.com/zorazrw/odex), and [SWE-bench](https://github.com/princeton-nlp/SWE-bench) 💔).

Reviewer 3: Maybe not to classify SWE-Bench into the "hard-to-deploy" category -- a more fundamental difference is probably many long-context or agent-oriented benchmarks are evaluating a compound subject -- mixing foundation models and the external feedback (e.g., RAG/code executor) -- while ours focuses on studying the raw intelligence and capability of the foundation model without external assistance. :)

The community still lacks an easy-to-use benchmark that can fundamentally evaluate the programming capabilities of LLMs.

> To address aforementioned limitations, we spent a whole year developing a new benchmark from scratch with the BigCode community. We are excited to announce the release of BigCodeBench, the benchmark that evaluates LLMs on solving practical and challenging programming tasks. Specifically, BigCodeBench contain 1,140 function-level tasks to challenge LLMs to follow instruction and compose multiple function calls as tools from 139 libraries. To evaluate LLMs rigorously, each programming task encompasses 5.6 test cases with an average branch coverage of 99%.

Author: I think we can remove a bit of the context and save it for the X thread? Worth highlighting the great effort in the thread.

Ready to deep dive into BigCodeBench? Let's get started! 🚀

## What does the task in BigCodeBench look like? 🕵️‍♂️

> <img src="https://github.com/bigcode-bench/bigcode-bench.github.io/blob/main/asset/tease.png?raw=true" alt="png" style="display: block; margin-left: auto; margin-right: auto;">

Reviewer 3: Consider exporting the Google drawing to SVG format to improve the image quality.

<img src="https://github.com/bigcode-bench/bigcode-bench.github.io/blob/main/asset/bigcodebench_prompt.png?raw=true" alt="png" style="display: block; margin-left: auto; margin-right: auto;">

As we can see from the task figure, the main target scenario is code completion (denoted as BigCodeBench-Complete), where LLMs are required to finish the implementation of the function based on the verbose instruction inside the docstring. However, considering the downstream applications such as multi-turn dialogue, users may tend to describe the requirements in a more conversational and less verbose manner. Instruction-tuned LLMs here come in handy, as they are trained to follow the natural-language-oriented instructions and generate code snippets accordingly. To test if the models are really capable enough to understand human intents to code, we create BigCodeBench-Instruct, a more challenging variant of BigCodeBench to particularly evaluate instruction-tuned LLMs.

## Where does the task come from? 🤔

<img src="https://github.com/bigcode-bench/bigcode-bench.github.io/blob/main/asset/construct_pipeline.png?raw=true" alt="png" style="display: block; margin-left: auto; margin-right: auto;">

> We guarantee the quality of the tasks in BigCodeBench by a systematic Human-LLM collaboration process. First, we choose [ODEX](https://github.com/zorazrw/odex) as the seed dataset, which contains short but realistic human intents and corresponding Python one-liners from Stack Overflow. We utilize GPT-4 to enrich the one-liners into comprehensive function-level tasks. Then, 20 human experts where most of them hold PhD degrees and 5+ years of Python programming experience voluntarily ground GPT-4 in the execution-based sandbox and continually instruct it to refactor the synthesized tasks and add test cases. Finally, we manually examine the tasks and test cases in the local environment, pre-evaluate the tasks on the other LLMs, and cross-check the tasks with 7 additional human experts to ensure the quality of the tasks. We sample tasks for 11 human experts to solve, and the average human performance is 97%.

Reviewer 2: ```suggestion
We guarantee the quality of the tasks in BigCodeBench by using a systematic "Human-LLM collaboration process". 
We use [ODEX](https://github.com/zorazrw/odex) as "seed dataset": it contains short but realistic human intents and corresponding Python one-liners from Stack Overflow, and we utilize GPT-4 to enrich the one-liners into comprehensive function-level tasks. Then, 20 human experts (where most of them hold PhD degrees and 5+ years of Python programming experience) voluntarily ground GPT-4 in an execution-based sandbox, and continually instruct it to refactor the synthesized tasks and add test cases. Finally, the tasks and test cases are examinated in the local environment, the tasks are pre-evaluated on other LLMs, and cross-checked with 7 additional human experts to ensure their quality. 

To assert overall quality, the authors sampled tasks for 11 human experts to solve, and the average human performance was 97%.
```

## How well do LLMs perform on BigCodeBench? 📊

> We host the BigCodeBench leaderboard on both [Hugging Face Space](https://huggingface.co/spaces/bigcode/bigcodebench-leaderboard) and [GitHub Pages](https://bigcode-bench.github.io/). The main difference between the two leaderboards is that the Hugging Face one is more interactive and provides more detailed information to analyze the performance of LLMs, while the GitHub Pages one is more stable and can be easily accessed by the community. So far, we have evaluated 71 LLMs across different sizes.

Reviewer 2: Not sure you need to explain differences between both

<script
	type="module"
	src="https://gradio.s3-us-west-2.amazonaws.com/4.36.1/gradio.js"
></script>

<gradio-app theme_mode="light" space="bigcode/bigcodebench-leaderboard"></gradio-app>

> We use Pass@1 with greedy decoding as the main evaluation metric on BigCodeBench. Interestingly, we observe that instruction-tuned LLMs like GPT-4 can omit the essential import statements of the given prompts

Reviewer 2: Metric description could go above in the tasks, with a small comment on how it works for the general community

> in BigCodeBench-Complete, which can lead to task failure due to the lack of proper module and constant definitions. Such behaviors are denoted as "model laziness" in long-context interactions, as [discussed in the community](https://community.openai.com/t/why-i-think-gpt-is-now-lazy/534332). To show the true capability of the models, we add the missing setup back during the Pass@1 evaluation, which is denoted as calibrated Pass@1. <u>Compared to human performance, we find that LLMs perform significantly lower on BigCodeBench-Complete, and even lower on BigCodeBench-Instruct, with the best model (GPT-4o) achieving the calibrated Pass@1 of 61.1% and 51.1%, respectively.</u> In addition, we observe a big performance gap between the close LLMs and open ones.

Reviewer 2: Calibrated pass@1 needs to be explained more

While Pass@1 is a good metric to show the overall performance of the models, it is not informative enough to do the pairwise comparison between the models. Inspired by [Chatbot Arena](https://lmsys.org/blog/2023-05-03-arena/), we employ Elo rating to rank the models on BigCodeBench-Complete. The Elo rating is a widely used method in chess to rank the players based on their performance in the games. We adapt the Elo rating to the programming tasks by treating each task as a game and each model as a player. The Elo rating is updated after each game based on the outcome of the game and the expected outcome of the game. We use the calibrated Pass@1 as the outcome of the game and exclude the tie results. We set the initial Elo rating of the models to 1000, fit it using maximum likelihood estimation, and bootstrap with 500 iterations to get the final scores. <u>We see that GPT-4o outperforms the other models by a large margin, and LLMs like Gemini-1.5-Pro, GPT-4-Turbo, GPT-4, Claude-3-Opus are in the second tier.</u>

> To help the community better understand how the models perform on each task, we provide the progress of solve rate, measured by calibrated Pass@1. On BigCodeBench-Complete, 151 tasks have not been solved by by all 71 models and 6 have been completely solved. For BigCodeBench-Instruct, there are 281 tasks remain unsolved and 14 fully solved by all models. As there is a significant number of unsolved tasks and only small number of fully solved tasks, we deem BigCodeBench as a challenging benchmark for LLMs.

Reviewer 2: This is super interesting!

To facilitate the evaluation, we provide pre-built Docker images for _code generation_ with [cuda-11.8.0](https://hub.docker.com/r/terryzho/bigcodebench-generate-cu11) and [cuda-12.1.1](https://hub.docker.com/r/terryzho/bigcodebench-generate-cu12), and [code execution](https://hub.docker.com/r/terryzho/bigcodebench-evaluate).

### Setup
```bash
# Install to use bigcodebench.evaluate
pip install bigcodebench --upgrade
# If you want to use the evaluate locally, you need to install the requirements
pip install -I -r https://raw.githubusercontent.com/bigcode-project/bigcodebench/main/Requirements/requirements-eval.txt

# Install to use bigcodebench.generate
# You are strongly recommended to install the [generate] dependencies in a separate environment
pip install bigcodebench[generate] --upgrade
```

> </details>

Reviewer 3: broken here.

### Code Generation
You are suggested to use `flash-attn` for generating code samples.
```bash
pip install -U flash-attn
```

To generate code samples from a model, you can use the following command:

> >

Reviewer 2: Remove the > as they mess up the markdown, it would seem

    --tp [gpu_number] \
    [--trust_remote_code] \
    [--base_url [base_url]]
```
>
The generated code samples will be stored in a file named `[model_name]--bigcodebench-[instruct|complete]--[backend]-[temp]-[n_samples].jsonl`. Alternatively, you can use the following command to utilize our pre-built docker images for generating code samples:
>
```shell
# If you are using GPUs
docker run --gpus '"device=$CUDA_VISIBLE_DEVICES"' -v $(pwd):/app -t terryzho/bigcodebench-generate-cu11:latest \
    --model [model_name] \ 
    --subset [complete|instruct] \
    [--greedy] \
    --bs [bs] \   
    --temperature [temp] \
    --n_samples [n_samples] \
    --resume \
    --backend [vllm|hf|openai|mistral|anthropic|google] \
    --tp [gpu_number]

# ...Or if you are using CPUs
docker run -v $(pwd):/app -t terryzho/bigcodebench-generate-cu11:latest \
    --model [model_name] \ 
    --subset [complete|instruct] \
    [--greedy] \
    --bs [bs] \   
    --temperature [temp] \
    --n_samples [n_samples] \
    --resume \
    --backend [vllm|hf|openai|mistral|anthropic|google]
```

>
```shell
# If you wish to use gated or private HuggingFace models and datasets

> docker run -e HUGGING_FACE_HUB_TOKEN=$token -v $(pwd):/app -t terryzho/bigcodebench-generate-cu11:latest # omit other arguments4

Reviewer 2: Why is there a user name here?

Author: I released the docker images under my account `terryzho`. Do you think it is better if I push the images to sth like `bigcodebench`?

Reviewer 2: Up to you, just wanted to make sure you were OK with this

Author: I pushed the images to `bigcodebench`, which looks better :)

```bash
# mount the current directory to the container
docker run -v $(pwd):/app terryzho/bigcodebench-evaluate:latest --subset [complete|instruct] --samples samples.jsonl
# ...Or locally ⚠️
bigcodebench.evaluate --subset [complete|instruct] --samples samples.jsonl
# ...If the ground truth is working locally
bigcodebench.evaluate --subset [complete|instruct] --samples samples.jsonl --no-gt
```

...Or if you want to try it locally regardless of the risks ⚠️:

First, install the dependencies for BigCodeBench:

```shell
pip install -r https://raw.githubusercontent.com/bigcode-project/bigcodebench/main/Requirements/requirements-eval.txt
```

Then, run the evaluation:

```bash
# ...Or locally ⚠️
bigcodebench.evaluate --subset [complete|instruct] --samples samples-calibrated.jsonl
# ...If the ground truth is not working locally
bigcodebench.evaluate --subset [complete|instruct] --samples samples-calibrated.jsonl --no-gt
```

### Failure Inspection

You can inspect the failed samples by using the following command:

```shell
bigcodebench.inspect --eval-results sample-sanitized_eval_results.json --in-place
```

> ## What's next?

Reviewer 2: I'll come back to this section tomorrow once the formatting issue is fixed
